Implement SHA-1 support in jest-haste-map

mjesun opened the pull request:
Adds support to `jest-haste-map` to provide in their public API the SHA-1 of the file. This SHA-1 is extracted from `watchman` when possible, and when not, it is processed by the worker tied to `jest-haste-map`. This however only happens when requested via the `sha1` option when building the Haste Map.

The benefits of this are mainly related to cache invalidation, which is the same reason on why it first got built in `watchman` itself.

I've added tests to confirm that when the SHA-1 is `null`, and it is requested by the user,  the worker will process it (which is what happens with the `node` crawler).

cpojer commented:
Super duper awesome. I'll take a better look tomorrow. What's the perf look like on www?

mjesun commented:
What is `www`? 😜

I haven't measured anywhere yet. But it should have no impact, since `sha1` is set to `false` by default. I can _try_ doing a follow-up change tomorrow to use that instead of opening and hashing the file in Jest.

cpojer commented on packages/jest-haste-map/src/__tests__/index.test.js:
```diff
@@ -8,6 +8,15 @@
 
+const crypto = require('crypto');
+
+function mockHashContents(contents) {
```
seems like const + arrow function allows you to oneline this guy.

mjesun replied:
The linter doesn't allow this 😢

And to have it always as multiline, I prefer using a proper function.

cpojer replied:
You mean the Jest hoist plugin? What about just inlining it in the mock below?

cpojer commented on packages/jest-haste-map/src/crawlers/watchman.js:
```diff
@@ -28,15 +28,20 @@ function WatchmanError(error: Error): Error {
   const client = new watchman.Client();
+
+  if (options.sha1) {
+    fields.push('content.sha1hex');
```
Don't we need to do a capability check with watchman? We need to stay backwards compatible here. To test, you can probably just use any property that doesn't exist and  see how watchman behaves. Watchman also has a capability check function to test whether this will work or not, we should be safe and fall back and calculate the sha1 separately like we do for the node crawler if that happens.

jeanlauliac replied:
Maybe the field specification is already ignored if it doesn't exist on watchman side? For robustness, I'd expect it to crash though, so I agree with Christoph a check (either capabilities, or perhaps just version number) is necessary to ensure we stay backward compatible.

We probably want to stay backward compatible because some companies may use older versions of Watchman with React Native, and may not be able to upgrade readily (for unrelated concerns, ex. compatibility with their source control).

mjesun replied:
I added a specific check for that by using `list-capabilities`.

cpojer commented on packages/jest-haste-map/src/index.js:
```diff
@@ -429,21 +437,28 @@ class HasteMap extends EventEmitter {
 
     const fileMetadata = hasteMap.files[filePath];
     const moduleMetadata = hasteMap.map[fileMetadata[H.ID]];
+    const sha1 = this._options.sha1 && !fileMetadata[H.SHA1];
```
This is a boolean, but the variable is called "sha1". This may be more readable if it was called `shouldComputeSha1` or similar.

jeanlauliac commented on packages/jest-haste-map/src/worker.js:
```diff
@@ -67,5 +70,17 @@ export async function worker(data: WorkerMessage): Promise<WorkerMetadata> {
 
-  return {dependencies, id, module};
+  // If a SHA-1 is requested on update, compute it.
+  if (data.sha1) {
```
Just to confirm: this will be falsy if watchman already provided the sha1? So it's a combination of "sha1 not provided by crawler/watcher" and "sha1 requested in jest-haste-map options"?

mjesun replied:
That is correct. You can see the combination of both parameters [here](https://github.com/facebook/jest/pull/5834/files/7157eec72ce0c503cf6f5ef8f53d29d526a8e8f6#diff-16f69297ff5bdc25a7a58f3ee8139521R440), on the main process.

mjesun commented:
I've pushed two new commits:

* One that renames all `sha1` booleans to `computeSha1`. However, I did not rename the `H.SHA1` constant, since this one _is_ actually a SHA-1 field.
* I added some logic into the watchman crawler to verify that the instance used supports `content.sha1hex`. If not, the field is not added to the list. I also added tests to verify that's the case.

cc / @jeanlauliac @cpojer

cpojer commented:
I'm good with this :) Nice work!

mjesun commented:
Let's wait for all CI to pass and I'll merge 🙂
